### PR TITLE
twister: verify platform names from quarantine

### LINKED ISSUE #43450: twister: platform names from quarantine file are not verified
**Describe the bug**
When quarantine.yaml file looks like this (with wrong platform name):
```
- scenarios:
    - sample.basic.helloworld
  platforms:
    - xyz
```

And Twister is called by following command:
```
./scripts/twister -p native_posix -T samples/hello_world/ --quarantine-list ~/zephyrproject/zephyr/quarantine.yaml
```

Program do not show any message, that platform from `quarantine.yaml` do not exist.

**To Reproduce**
1. Create following `quarantine.yaml` file:
```
- scenarios:
    - sample.basic.helloworld
  platforms:
    - xyz
```
2. Run following command:
```
./scripts/twister -p native_posix -T samples/hello_world/ --quarantine-list ~/zephyrproject/zephyr/quarantine.yaml
```
3. Observe, that Twister do not detect that given platform name do not exist.

**Expected behavior**
Twister should show message, that `xyz` platform from `quarantine.yaml` file do not exist.

**Impact**
In specific situations typo in `quarantine.yaml` platform list can be silently skipped.

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK
 - Commit SHA: a036d256a998ff0c37162f7bf6fe2bb4d4d7cf57

**Additional context**
I will crate PR with suitable fix soon.

### PR DESCRIPTION
**Pull request by gopiotr**
Add verification of platform name from quarantine list.

Fixes: #43450

Signed-off-by: Piotr Golyzniak <piotr.golyzniak@nordicsemi.no>

### CODE REVIEW

File: scripts/pylib/twister/twisterlib.py [hunk 1]
```diff
@@ -3163,13 +3163,17 @@ def load_quarantine(self, file):
         # platforms and scenarios for each entry in quarantine yaml
         quarantine_list = []
         for quar_dict in quarantine_yaml:
+            scenarios = quar_dict['scenarios']
+            scenarios = [s for s in scenarios if s is not None]
             if quar_dict['platforms'][0] == "all":
                 plat = self.platform_names
             else:
                 plat = quar_dict['platforms']
+                self.verify_platforms_existence(plat, f"quarantine-list")
```
Comment (chen-png) [inline scripts/pylib/twister/twisterlib.py:3172]: this function was invoked in other place too, but you add a new support for platform with none and only filter it in this place, you should handle those place too.
or I think you can filter none platforms firstly like line 3173 , then invoke verify function, in this case, don't need to modify verify_platforms_existence function.

Comment (gopiotr) [reply scripts/pylib/twister/twisterlib.py:3172]: Originally, I added "support" for `null` platform in `verify_platforms_existence(...)` function at first to be able to properly handle situation in yaml file, when somebody would like to enforce no execution with `--integration` mode:
```
integration_platforms:
  - null
```
Then when I wanted to apply the same "handling" to verify platform names from quarantine file I observed that `null` platform is converted into None object which cannot be used with `join` function in 3175 line and cause error. That's why I added filtering of `null` platform in 3173 line.

I think that this is better approach, because user is still informed in debug message, that `null` platform is detected (in `verify_platforms_existence(...)` function) and in the same time it do not require to change parsing of each platform lists and removing every `null` platform (only in quarantine case).

**Review by enjiamai — DISMISSED** [review]

**Review by PerMac — DISMISSED** [review]

**Review by nashif — CHANGES_REQUESTED** [review]
The semantics of 

```
integration_platforms:
  - null
``` 

is not clear to me, it is either you have no integration_platforms, or you have them, why are we introducing this syntax?

**Comment by PerMac** [comment]
TL;DR: This could help with test scope traceability and is required if one decides to use quarantine in their CI. 

1. One can have a file with tests under quarantine in their repo, which is used when running their CI. Since quarantine should only be temporary, the file should be "empty", however this causes issues. The solution is to keep quarantine file with the scope of scenarios and platforms set as `None` (e.g. https://github.com/nrfconnect/sdk-nrf/blob/main/scripts/quarantine.yaml). From some time twister is verifying if the platform name exist in the platforms definitions and it complains that `None` is not a proper platform name. 

2. As for integration_platforms: At nrfconnect we were able to add `integration_platforms` section to all our samples and tests. There are only a few special cases, where maintainers decided they don't want their tests to be run in the CI at all (e.g. they have special CIs for them). With being able to have `integration_platforms: -null` we will be able to add automatic verification checking if every test/sample in our repo has `integraton_platforms` entry added. Without it we have to manually verify this. 
We found using `integration_platforms` as the only way to enforce reliable scope for tests in the CI and we made a policy, that by default all testcase/sample.yamls must have `integration_platforms`. Setting it to `null` is a way to allow for exceptions wihout twister complaining that it doesn't find it as a valid platform.

Explanation of point 2. relates to nrfconnect repo, however the approach we took could also be used in the upstream. IMO defining `integration_platforms` scope for every yaml would be beneficial for the transparency and reliability of the upstream CI.

**Comment by PerMac** [comment]
@nashif Does my explanation makes sense? Can you please rethink this having the above in mind?

**Comment by nashif** [comment]
> 1. From some time twister is verifying if the platform name exist in the platforms definitions and it complains that `None` is not a proper platform name.

why use None? Why dont you just use an empty list...

```
platforms: []
```

> Edit: 3. In addition, running integration_mode outside of the upstream context is problematic without having integration platforms explicitly defined in every yaml. Without it twister defines the scope by its own, which is very confusing for the developers, and in most cases doesn't make sense outside of the upstream repo (Twister can decide to use its "default platforms").

You can tell twister which platforms you want to target using --platform. Twister does not decide on its own, this is a wrong statement. Twister decides based on configuration.



> 2\. As for integration_platforms: At nrfconnect we were able to add `integration_platforms` section to all our samples and tests. There are only a few special cases, where maintainers decided they don't want their tests to be run in the CI at all (e.g. they have special CIs for them). With being able to have `integration_platforms: -null` we will be able to add automatic verification checking if every test/sample in our repo has `integraton_platforms` entry added. Without it we have to manually verify this.
>     We found using `integration_platforms` as the only way to enforce reliable scope for tests in the CI and we made a policy, that by default all testcase/sample.yamls must have `integration_platforms`. Setting it to `null` is a way to allow for exceptions without twister complaining that it doesn't find it as a valid platform.

This looks like using integration-platforms to declare your own set of default platforms. Seems very custom and complicating things unnecessarily. Summary: ```integration-platforms``` defined as a null is a no-no from me.

We need to look into some kind of local configuration support that would override anything coming from the code directly (upstream code), i.e. defining your own "default" platforms, what tags to ignore and things defined in platform yaml file in addition to maybe integrating the functionality of quarantine into that as well. One single local configuration file that can be used to define a different test environment from the one we use mainly to manage CI.

**Comment by gopiotr** [comment]
@nashif @PerMac @chen-png 

To make it possible to close issue https://github.com/zephyrproject-rtos/zephyr/issues/43450 I decided to leave only this part connected with adding verification of platforms list from quarantine file. We can return to discussion about passing `null` platform in `integration_platforms` in the future.

**Review by hakehuang — APPROVED** [review]

**Review by chen-png — APPROVED** [review]

**Review by PerMac — APPROVED** [review]

**Review by nashif — APPROVED** [review]
